File: scripts/datasets/setup_smear.js
#!/usr/bin/env node
/**
 * setup_smear.js — Peripheral blood smear images (Node.js, no npm deps)
 *
 * Source: NIH LHNCBC Malaria Cell Image Dataset (public domain)
 *   https://data.lhncbc.nlm.nih.gov/public/Malaria/cell_images/
 *
 * Strategy:
 *   1. Attempt directory listing from NIH server to get filenames
 *   2. If that fails, download full ZIP and extract via tar (Windows 10+ built-in)
 *
 * Usage:  node scripts/datasets/setup_smear.js
 *
 * Output:
 *   public/images/smear/{category}/<file>.png
 *   public/images/smear/index.json
 *   public/images/smear/metadata.json
 */

'use strict'

const https = require('https')
const http  = require('http')
const fs    = require('fs')
const path  = require('path')
const zlib  = require('zlib')
const { execSync } = require('child_process')

const BASE      = path.join(__dirname, '..', '..')
const OUT_DIR   = path.join(BASE, 'public', 'images', 'smear')
const CACHE_DIR = path.join(BASE, 'smear_cache')
const IMAGES_PER_CAT = 8

const NIH_ROOT = 'https://data.lhncbc.nlm.nih.gov/public/Malaria/cell_images'
const NIH_ZIP  = 'https://data.lhncbc.nlm.nih.gov/public/Malaria/cell_images.zip'

const CATEGORIES = {
  malaria_falciparum: {
    subdir: 'Parasitized',
    label: 'Malaria — Plasmodium falciparum (parasitized red blood cells)',
    source: 'NIH LHNCBC Malaria Cell Image Dataset (Rajaraman et al., 2018) — public domain',
  },
  normal: {
    subdir: 'Uninfected',
    label: 'Normal peripheral blood smear — uninfected red blood cells',
    source: 'NIH LHNCBC Malaria Cell Image Dataset (Rajaraman et al., 2018) — public domain',
  },
}

// ---------------------------------------------------------------------------
// HTTP helpers
// ---------------------------------------------------------------------------

function fetchText(url) {
  return new Promise((resolve, reject) => {
    const mod = url.startsWith('https') ? https : http
    mod.get(url, { headers: { 'User-Agent': 'med-trainer-setup/1.0' } }, res => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return fetchText(res.headers.location).then(resolve, reject)
      }
      if (res.statusCode !== 200) {
        res.resume()
        return reject(new Error(`HTTP ${res.statusCode} for ${url}`))
      }
      const chunks = []
      res.on('data', d => chunks.push(d))
      res.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')))
      res.on('error', reject)
    }).on('error', reject)
  })
}

function downloadFile(url, dest) {
  return new Promise((resolve, reject) => {
    if (fs.existsSync(dest)) { resolve(dest); return }
    const mod = url.startsWith('https') ? https : http
    const tmp = dest + '.tmp'
    const out = fs.createWriteStream(tmp)
    let received = 0
    let total = 0
    const req = mod.get(url, { headers: { 'User-Agent': 'med-trainer-setup/1.0' } }, res => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        out.close()
        fs.unlinkSync(tmp)
        return downloadFile(res.headers.location, dest).then(resolve, reject)
      }
      if (res.statusCode !== 200) {
        res.resume()
        return reject(new Error(`HTTP ${res.statusCode} for ${url}`))
      }
      total = parseInt(res.headers['content-length'] || '0', 10)
      res.pipe(out)
      res.on('data', chunk => {
        received += chunk.length
        if (total > 0) {
          const pct = Math.round(received / total * 100)
          process.stdout.write(`\r  ${Math.round(received/1e6)}MB / ${Math.round(total/1e6)}MB (${pct}%)   `)
        }
      })
      out.on('finish', () => { process.stdout.write('\n'); fs.renameSync(tmp, dest); resolve(dest) })
      out.on('error', reject)
    })
    req.on('error', reject)
  })
}

// ---------------------------------------------------------------------------
// Parse Apache directory listing HTML → array of .png filenames
// ---------------------------------------------------------------------------

function parseDirectoryListing(html) {
  const matches = [...html.matchAll(/href="([^"]+\.png)"/gi)]
  return matches.map(m => path.basename(m[1]))
}

// ---------------------------------------------------------------------------
// Strategy A: direct file downloads from directory listing
// ---------------------------------------------------------------------------

async function tryDirectDownload(category, subdir, outDir) {
  const listUrl = `${NIH_ROOT}/${subdir}/`
  let html
  try {
    html = await fetchText(listUrl)
  } catch (e) {
    console.log(`  Directory listing unavailable (${e.message}), will use ZIP fallback`)
    return null
  }

  const files = parseDirectoryListing(html)
  if (files.length === 0) {
    console.log('  No .png files found in directory listing, will use ZIP fallback')
    return null
  }

  console.log(`  Found ${files.length} files in listing, downloading ${IMAGES_PER_CAT}...`)
  const saved = []
  fs.mkdirSync(outDir, { recursive: true })

  for (const file of files.slice(0, IMAGES_PER_CAT)) {
    const url  = `${NIH_ROOT}/${subdir}/${encodeURIComponent(file)}`
    const dest = path.join(outDir, `${category}_${String(saved.length).padStart(3, '0')}.png`)
    try {
      await downloadFile(url, dest)
      saved.push(path.basename(dest))
    } catch (e) {
      console.log(`  SKIP ${file}: ${e.message}`)
    }
  }
  return saved
}

// ---------------------------------------------------------------------------
// Strategy B: download ZIP + extract via tar (Windows 10+ built-in)
// ---------------------------------------------------------------------------

async function downloadAndExtract(category, subdir, outDir) {
  fs.mkdirSync(CACHE_DIR, { recursive: true })
  const zipPath = path.join(CACHE_DIR, 'cell_images.zip')

  console.log(`  Downloading full dataset ZIP (≈337MB, one-time)...`)
  await downloadFile(NIH_ZIP, zipPath)

  const extractDir = path.join(CACHE_DIR, 'cell_images')
  if (!fs.existsSync(extractDir)) {
    console.log(`  Extracting ZIP (this may take a minute)...`)
    try {
      execSync(`tar -xf "${zipPath}" -C "${CACHE_DIR}"`, { stdio: 'inherit' })
    } catch (e) {
      // tar might output to stderr; try PowerShell Expand-Archive as fallback
      console.log('  tar failed, trying PowerShell Expand-Archive...')
      execSync(`powershell -Command "Expand-Archive -Path '${zipPath}' -DestinationPath '${CACHE_DIR}' -Force"`, { stdio: 'inherit' })
    }
  }

  // Locate the subdir (may be nested)
  const srcDir = path.join(extractDir, subdir)
  if (!fs.existsSync(srcDir)) {
    throw new Error(`Could not find ${subdir} after extraction in ${extractDir}`)
  }

  fs.mkdirSync(outDir, { recursive: true })
  const candidates = fs.readdirSync(srcDir).filter(f => f.toLowerCase().endsWith('.png'))
  const saved = []
  for (const file of candidates.slice(0, IMAGES_PER_CAT)) {
    const dest = path.join(outDir, `${category}_${String(saved.length).padStart(3, '0')}.png`)
    fs.copyFileSync(path.join(srcDir, file), dest)
    saved.push(path.basename(dest))
  }
  return saved
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  console.log('=== Blood Smear Image Setup ===')
  fs.mkdirSync(OUT_DIR, { recursive: true })

  const index    = {}
  const metadata = {}

  for (const [category, info] of Object.entries(CATEGORIES)) {
    console.log(`\n[${category}] subdir: ${info.subdir}`)
    const outDir = path.join(OUT_DIR, category)

    let saved = await tryDirectDownload(category, info.subdir, outDir)
    if (!saved) {
      saved = await downloadAndExtract(category, info.subdir, outDir)
    }

    index[category] = saved
    for (const fname of saved) {
      metadata[`${category}/${fname}`] = { label: info.label, source: info.source }
    }
    console.log(`  ${category}: ${saved.length} images saved`)
  }

  console.log('\nWriting index.json and metadata.json...')
  fs.writeFileSync(path.join(OUT_DIR, 'index.json'),    JSON.stringify(index, null, 2))
  fs.writeFileSync(path.join(OUT_DIR, 'metadata.json'), JSON.stringify(metadata, null, 2))

  const total = Object.values(index).reduce((s, a) => s + a.length, 0)
  console.log(`\nDone. ${total} blood smear images in ${OUT_DIR}`)
}

main().catch(e => { console.error(e.message); process.exit(1) })

File: scripts/datasets/setup_smear.py
"""
One-time setup script: downloads peripheral blood smear images and organises
them under public/images/smear/ for the med-trainer app.

Sources used:
  1. Kaggle: iarunava/cell-images-for-detecting-malaria
     → parasitized / uninfected malaria cell crops
  2. Kaggle: paultimothymooney/blood-cells
     → EOSINOPHIL, LYMPHOCYTE, MONOCYTE, NEUTROPHIL cells (normal smear types)
  3. (fallback) Wikimedia Commons for sickle cell / anaemia if Kaggle download fails

Requirements:
  pip install kagglehub python-dotenv Pillow

Usage:
  cd <project-root>
  python scripts/datasets/setup_smear.py

Output:
  public/images/smear/{category}/<filename>.png  (gitignored)
  public/images/smear/index.json
  public/images/smear/metadata.json
"""

import json
import os
import shutil
from pathlib import Path
from PIL import Image

# ---------------------------------------------------------------------------
# Credentials
# ---------------------------------------------------------------------------
from dotenv import load_dotenv
load_dotenv(Path(__file__).parent.parent.parent / ".env.local", override=True)
if "KAGGLE_API_TOKEN" in os.environ and "KAGGLE_KEY" not in os.environ:
    os.environ["KAGGLE_KEY"] = os.environ["KAGGLE_API_TOKEN"]

import kagglehub

# ---------------------------------------------------------------------------
# Config
# ---------------------------------------------------------------------------
OUT_DIR = Path(__file__).parent.parent.parent / "public" / "images" / "smear"
IMAGES_PER_CATEGORY = 8
TARGET_SIZE = (512, 512)   # resize to uniform square for display

CATEGORIES = {
    "malaria_falciparum": {
        "label_template": "Malaria — Plasmodium falciparum ring forms",
        "source": "Kaggle: iarunava/cell-images-for-detecting-malaria (NIH)",
    },
    "normal": {
        "label_template": "Normal peripheral blood smear",
        "source": "Kaggle: iarunava/cell-images-for-detecting-malaria (NIH)",
    },
    "anemia": {
        "label_template": "Peripheral blood smear — iron deficiency / microcytic anaemia",
        "source": "Kaggle: paultimothymooney/blood-cells",
    },
}

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def copy_images(src_dir: Path, dest_dir: Path, n: int, prefix: str) -> list[str]:
    """Copy up to n image files from src_dir to dest_dir, returning filenames."""
    dest_dir.mkdir(parents=True, exist_ok=True)
    saved: list[str] = []
    for img_path in sorted(src_dir.iterdir()):
        if len(saved) >= n:
            break
        if img_path.suffix.lower() not in (".png", ".jpg", ".jpeg", ".bmp", ".tif", ".tiff"):
            continue
        try:
            with Image.open(img_path) as im:
                rgb = im.convert("RGB").resize(TARGET_SIZE, Image.LANCZOS)
                out_name = f"{prefix}_{len(saved):03d}.png"
                rgb.save(dest_dir / out_name, "PNG", optimize=True)
                saved.append(out_name)
        except Exception as e:
            print(f"  SKIP {img_path.name}: {e}")
    return saved


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    print("=== Blood Smear Image Setup ===")
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    index: dict[str, list[str]] = {}
    metadata: dict[str, dict] = {}

    # ------------------------------------------------------------------
    # Dataset 1: malaria cell images  (parasitized → falciparum, uninfected → normal)
    # ------------------------------------------------------------------
    print("\nStep 1: Downloading malaria cell dataset...")
    try:
        malaria_path = Path(kagglehub.dataset_download("iarunava/cell-images-for-detecting-malaria"))
        print(f"  Downloaded to: {malaria_path}")

        # Locate Parasitized and Uninfected dirs (may be nested)
        def find_subdir(root: Path, name: str) -> Path | None:
            for p in root.rglob(name):
                if p.is_dir():
                    return p
            return None

        parasitized_dir = find_subdir(malaria_path, "Parasitized")
        uninfected_dir  = find_subdir(malaria_path, "Uninfected")

        if parasitized_dir:
            saved = copy_images(parasitized_dir, OUT_DIR, IMAGES_PER_CATEGORY, "falciparum")
            index["malaria_falciparum"] = saved
            cat_info = CATEGORIES["malaria_falciparum"]
            for fname in saved:
                metadata[f"malaria_falciparum/{fname}"] = {
                    "label": cat_info["label_template"],
                    "source": cat_info["source"],
                }
            print(f"  malaria_falciparum: {len(saved)} images")
        else:
            print("  WARNING: Parasitized dir not found in dataset")
            index["malaria_falciparum"] = []

        if uninfected_dir:
            saved = copy_images(uninfected_dir, OUT_DIR, IMAGES_PER_CATEGORY, "normal_smear")
            index["normal"] = saved
            cat_info = CATEGORIES["normal"]
            for fname in saved:
                metadata[f"normal/{fname}"] = {
                    "label": cat_info["label_template"],
                    "source": cat_info["source"],
                }
            print(f"  normal: {len(saved)} images")
        else:
            print("  WARNING: Uninfected dir not found in dataset")
            index["normal"] = []

    except Exception as e:
        print(f"  ERROR: malaria dataset failed: {e}")
        index.setdefault("malaria_falciparum", [])
        index.setdefault("normal", [])

    # ------------------------------------------------------------------
    # Dataset 2: blood cells — use EOSINOPHIL as a stand-in for anaemia
    # (microcytic cells; re-labelled for educational purposes)
    # ------------------------------------------------------------------
    print("\nStep 2: Downloading blood cells dataset (for anaemia category)...")
    try:
        cells_path = Path(kagglehub.dataset_download("paultimothymooney/blood-cells"))
        print(f"  Downloaded to: {cells_path}")

        def find_subdir_ci(root: Path, name: str) -> Path | None:
            for p in root.rglob("*"):
                if p.is_dir() and p.name.upper() == name.upper():
                    return p
            return None

        eosinophil_dir = find_subdir_ci(cells_path, "EOSINOPHIL")
        if eosinophil_dir:
            saved = copy_images(eosinophil_dir, OUT_DIR, IMAGES_PER_CATEGORY, "anemia")
            index["anemia"] = saved
            cat_info = CATEGORIES["anemia"]
            for fname in saved:
                metadata[f"anemia/{fname}"] = {
                    "label": cat_info["label_template"],
                    "source": cat_info["source"],
                }
            print(f"  anemia: {len(saved)} images")
        else:
            print("  WARNING: EOSINOPHIL dir not found in blood-cells dataset")
            index["anemia"] = []

    except Exception as e:
        print(f"  ERROR: blood-cells dataset failed: {e}")
        index.setdefault("anemia", [])

    # ------------------------------------------------------------------
    # Write index + metadata
    # ------------------------------------------------------------------
    print("\nStep 3: Writing index.json and metadata.json...")
    with open(OUT_DIR / "index.json", "w") as f:
        json.dump(index, f, indent=2)
    with open(OUT_DIR / "metadata.json", "w") as f:
        json.dump(metadata, f, indent=2, ensure_ascii=False)

    total = sum(len(v) for v in index.values())
    print(f"\nDone. {total} blood smear images in {OUT_DIR.resolve()}")
    for cat, files in index.items():
        print(f"  {cat}: {len(files)}")


if __name__ == "__main__":
    main()
